new upsync type consul_services

lyokha opened the pull request:
This commit makes possible to get upstreams from the Consul services catalog. For testing I used the following Consul config:

```json
{
    "server": true,
    "node_name": "test",
    "bind_addr": "127.0.0.1",
    "data_dir": "/var/lib/consul",
    "log_level": "INFO",
    "service": {"name": "test",
                "port": 8081}
}
```

and Nginx config (it requires nginx *echo* module), notice that `upsync_type` is `consul_services`:

```nginx
user                    nobody;
worker_processes        2;

events {
    worker_connections  1024;
}

http {
    upstream test {
        # fake server otherwise ngx_http_upstream will report error when startup
        server 127.0.0.1:11111;

        # all backend server will pull from consul when startup and will delete fake server
        upsync 127.0.0.1:8500/v1/catalog/service/test
                upsync_timeout=6m
                upsync_interval=500ms
                upsync_type=consul_services
                strong_dependency=off;
        #upsync_dump_path /usr/local/nginx/conf/servers/servers_test.conf;
        upsync_dump_path /tmp/servers_test.conf;
    }

    server {
        listen 8080;

        location = /proxy_test {
            proxy_pass http://test;
        }

        location = /upstream_show {
            upstream_show;
        }
    }

    server {
        listen 8081;

        location / {
            echo "In 8081";
        }
    }
}
```

The arguable solutions are:

1. All servers have default hard-coded values for `max_fails`, `fail_timeout` and other server-specific parameters as there is no data available from services catalog to map to them. Some of the parameters (like `backup` or `down`) cannot be figured out at all, but `max_fails` and `fail_timeout` can be set statically on per-upstream basis using dedicated directives like `upsync_consul_services_max_fails` and `upsync_consul_services_fail_timeout` - they are not implemented here but it's easy to do.
2. If `NGX_HTTP_UPSYNC_CONSUL` and others are planned to be used in bit operations then their values must be `0x0001`, `0x0002` and `0x0004` instead of current 1 to 3. Also I put the new value `NGX_HTTP_UPSYNC_CONSUL_SERVICES` on the 2nd place (`0x0002`).
3. I have not changed the request URI for the services catalog and it includes parameter `recurse`: this is not needed by Consul but seems to be harmless.

gfrankliu commented:
Can we use service tags for those server specific parameters? e.g.:

```
{
    "server": true,
    "node_name": "test",
    "bind_addr": "127.0.0.1",
    "data_dir": "/var/lib/consul",
    "log_level": "INFO",
    "service": {"name": "test",
                       "port": 8081,
                       "tags": [
                                        "max_fails=1",
                                        "fail_timeout=10s",
                                        "weight=5",
                                        "backup"
                                       ],
                        "check" {
                               "TCP": "127.0.0.1:8081",
                               "Interval": "5s"
                       }
        }
}
```

gfrankliu commented:
Code can filter out those tags that are irrelevant to nginx.

lyokha commented:
Yes, this makes real sense and is not hard to implement.

xiaokai-wang commented:
It's welcome.  Thanks.

What do you think? @gfrankliu

gfrankliu commented:
@xiaokai-wang I definitely like this feature. I will wait until @lyokha adds the "tag" support before trying out.

lyokha commented:
I implemented this. My consul config I tested is

```json
{
    "server": true,
    "node_name": "test",
    "bind_addr": "127.0.0.1",
    "data_dir": "/var/lib/consul",
    "log_level": "INFO",
    "service": {"name": "test",
                "port": 8081,
                "tags": ["weight=4"
                        ,"max_fails=8"
                        ,"fail_timeout=10s"
                        ,"backup"]}
}
```

The nginx config is the same. When you add "down" to list of tags, you'll get "no live upstreams" in log and status 502 as expected.

The implementation requires that attributes with value have no space after their name and sign `=` (see in the patch).

I also needed to change json source file, but I did not include it in the commit, here it is:

```diff
diff --git a/src/ngx_http_json.c b/src/ngx_http_json.c
index ce950de..fdcf930 100644
--- a/src/ngx_http_json.c
+++ b/src/ngx_http_json.c
@@ -38,7 +38,8 @@ const char *cJSON_GetErrorPtr(void) {return ep;}
 
 static int cJSON_strcasecmp(const char *s1,const char *s2)
 {
-	if (!s1) return (s1==s2)?0:1;if (!s2) return 1;
+	if (!s1) return (s1==s2)?0:1;
+	if (!s2) return 1;
 	for(; tolower(*s1) == tolower(*s2); ++s1, ++s2)	if(*s1 == 0)	return 0;
 	return tolower(*(const unsigned char *)s1) - tolower(*(const unsigned char *)s2);
 }
@@ -484,7 +485,8 @@ static char *print_object(cJSON *item,int depth,int fmt)
 		*ptr++=':';if (fmt) *ptr++='\t';
 		strcpy(ptr,entries[i]);ptr+=strlen(entries[i]);
 		if (i!=numentries-1) *ptr++=',';
-		if (fmt) *ptr++='\n';*ptr=0;
+		if (fmt) *ptr++='\n';
+		*ptr=0;
 		cJSON_free(names[i]);cJSON_free(entries[i]);
 	}
 	
@@ -511,7 +513,10 @@ void	cJSON_AddItemReferenceToArray(cJSON *array, cJSON *item)						{cJSON_AddIte
 void	cJSON_AddItemReferenceToObject(cJSON *object,const char *string,cJSON *item)	{cJSON_AddItemToObject(object,string,create_reference(item));}
 
 cJSON *cJSON_DetachItemFromArray(cJSON *array,int which)			{cJSON *c=array->child;while (c && which>0) c=c->next,which--;if (!c) return 0;
-	if (c->prev) c->prev->next=c->next;if (c->next) c->next->prev=c->prev;if (c==array->child) array->child=c->next;c->prev=c->next=0;return c;}
+	if (c->prev) c->prev->next=c->next;
+	if (c->next) c->next->prev=c->prev;
+	if (c==array->child) array->child=c->next;
+	c->prev=c->next=0;return c;}
 void   cJSON_DeleteItemFromArray(cJSON *array,int which)			{cJSON_Delete(cJSON_DetachItemFromArray(array,which));}
 cJSON *cJSON_DetachItemFromObject(cJSON *object,const char *string) {int i=0;cJSON *c=object->child;while (c && cJSON_strcasecmp(c->string,string)) i++,c=c->next;if (c) return cJSON_DetachItemFromArray(object,i);return 0;}
 void   cJSON_DeleteItemFromObject(cJSON *object,const char *string) {cJSON_Delete(cJSON_DetachItemFromObject(object,string));}
```

Without it I was getting compilation errors on my Fedora 25 / gcc 6.3.1 / `-Werror` on:

```
02/04/17 12:05⮁⮁ lyokha@desktop⮁⮁ nginx-1.10.2 ⮀ make
/usr/bin/make -f objs/Makefile
make[1]: вход в каталог «/home/lyokha/Загрузки/nginx-1.10.2»
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused -Werror -g  -I/usr/lib64/ghc-8.0.1/include  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I /home/lyokha/devel/nginx-upsync-module/src \
	-o objs/addon/src/ngx_http_upsync_module.o \
	/home/lyokha/devel/nginx-upsync-module/src/ngx_http_upsync_module.c
cc -c -pipe  -O -W -Wall -Wpointer-arith -Wno-unused -Werror -g  -I/usr/lib64/ghc-8.0.1/include  -I src/core -I src/event -I src/event/modules -I src/os/unix -I objs -I src/http -I src/http/modules -I /home/lyokha/devel/nginx-upsync-module/src \
	-o objs/addon/src/ngx_http_json.o \
	/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c: В функции «cJSON_strcasecmp»:
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c:41:2: ошибка: this «if» clause does not guard... [-Werror=misleading-indentation]
  if (!s1) return (s1==s2)?0:1;if (!s2) return 1;
  ^~
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c:41:31: замечание: ...this statement, but the latter is misleadingly indented as if it is guarded by the «if»
  if (!s1) return (s1==s2)?0:1;if (!s2) return 1;
                               ^~
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c: В функции «print_object»:
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c:487:3: ошибка: this «if» clause does not guard... [-Werror=misleading-indentation]
   if (fmt) *ptr++='\n';*ptr=0;
   ^~
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c:487:24: замечание: ...this statement, but the latter is misleadingly indented as if it is guarded by the «if»
   if (fmt) *ptr++='\n';*ptr=0;
                        ^
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c: В функции «cJSON_DetachItemFromArray»:
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c:514:2: ошибка: this «if» clause does not guard... [-Werror=misleading-indentation]
  if (c->prev) c->prev->next=c->next;if (c->next) c->next->prev=c->prev;if (c==array->child) array->child=c->next;c->prev=c->next=0;return c;}
  ^~
/home/lyokha/devel/nginx-upsync-module/src/ngx_http_json.c:514:37: замечание: ...this statement, but the latter is misleadingly indented as if it is guarded by the «if»
  if (c->prev) c->prev->next=c->next;if (c->next) c->next->prev=c->prev;if (c==array->child) array->child=c->next;c->prev=c->next=0;return c;}
                                     ^~
cc1: все предупреждения считаются ошибками
objs/Makefile:1279: ошибка выполнения рецепта для цели «objs/addon/src/ngx_http_json.o»
make[1]: *** [objs/addon/src/ngx_http_json.o] Ошибка 1
make[1]: выход из каталога «/home/lyokha/Загрузки/nginx-1.10.2»
Makefile:8: ошибка выполнения рецепта для цели «build»
make: *** [build] Ошибка 2
```

(there is some Russian here due to my locale). The reason is simple: gcc don't like when after `if` stament without the block there are other statements on the same line.
Should I include this change too?

gfrankliu commented:
I assume the json error is due to gcc version and unrelated to this PR?

lyokha commented:
@gfrankliu Yes, this is unrelated.

gfrankliu commented:
I suggest you submit another PR for fixing the compiler -Werror errors with latest gcc.

lyokha commented:
Ok, agreed.

xiaokai-wang commented:
Tested and used in your company environment?

Having test case will be perfect.

lyokha commented:
We made very basic manual tests: adding/removing servers, so far it's Ok. But I do not know details as soon as the tests were run by another team. And we had slightly extended version of the patch: with possibility to completely remove all servers from the service list using so called *fallback peer* marked as *down*, I can show this additional patch if you want.

xiaokai-wang commented:
On the other hand, can you update readme? Thanks.

lyokha commented:
Updated (see consul interface). I also added a guard check in *ngx_http_upsync_consul_parse_json()* - checking if *upstream_conf* is not NULL saves a worker from segmentation fault when using services catalog link but having forgot to set *upsync_type* to *consul_services* instead of *consul*.

lyokha commented:
Here is additional patch for the extended version we tested (added directives *upsync_consul_services_max_fails*, *upsync_consul_services_fail_timeout* for setting default values just from upstream block, and *upsync_fallback_peer* to make it possible to delete all servers from an upstream: it is not currently feasible in upsync module).

```diff
diff --git a/src/ngx_http_upsync_module.c b/src/ngx_http_upsync_module.c
index 0ef5a08..79b3a6a 100644
--- a/src/ngx_http_upsync_module.c
+++ b/src/ngx_http_upsync_module.c
@@ -86,6 +86,11 @@ typedef struct {
     ngx_int_t                        upsync_lb;
     ngx_uint_t                       strong_dependency;
 
+    ngx_uint_t                       consul_services_max_fails;
+    time_t                           consul_services_fail_timeout;
+
+    ngx_http_upstream_rr_peer_t      fallback_peer;
+
     ngx_str_t                        upsync_send;
     ngx_str_t                        upsync_dump_path;
 
@@ -160,6 +165,12 @@ static char *ngx_http_upsync_server(ngx_conf_t *cf,
     ngx_command_t *cmd, void *conf);
 static char *ngx_http_upsync_set_conf_dump(ngx_conf_t *cf, 
     ngx_command_t *cmd, void *conf);
+static char *ngx_http_upsync_consul_services_max_fails(ngx_conf_t *cf, 
+    ngx_command_t *cmd, void *conf);
+static char *ngx_http_upsync_consul_services_fail_timeout(ngx_conf_t *cf, 
+    ngx_command_t *cmd, void *conf);
+static char *ngx_http_upsync_fallback_peer(ngx_conf_t *cf, 
+    ngx_command_t *cmd, void *conf);
 
 static void *ngx_http_upsync_create_main_conf(ngx_conf_t *cf);
 static void *ngx_http_upsync_create_srv_conf(ngx_conf_t *cf);
@@ -290,6 +301,27 @@ static ngx_command_t  ngx_http_upsync_commands[] = {
         0,
         NULL },
 
+    {  ngx_string("upsync_consul_services_max_fails"),
+        NGX_HTTP_UPS_CONF|NGX_CONF_TAKE1,
+        ngx_http_upsync_consul_services_max_fails,
+        NGX_HTTP_SRV_CONF_OFFSET,
+        0,
+        NULL },
+
+    {  ngx_string("upsync_consul_services_fail_timeout"),
+        NGX_HTTP_UPS_CONF|NGX_CONF_TAKE1,
+        ngx_http_upsync_consul_services_fail_timeout,
+        NGX_HTTP_SRV_CONF_OFFSET,
+        0,
+        NULL },
+
+    {  ngx_string("upsync_fallback_peer"),
+        NGX_HTTP_UPS_CONF|NGX_CONF_TAKE1,
+        ngx_http_upsync_fallback_peer,
+        NGX_HTTP_SRV_CONF_OFFSET,
+        0,
+        NULL },
+
     {  ngx_string("upstream_show"),
         NGX_HTTP_LOC_CONF|NGX_CONF_NOARGS,
         ngx_http_upsync_set,
@@ -639,6 +671,76 @@ ngx_http_upsync_set_conf_dump(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 }
 
 
+static char *
+ngx_http_upsync_consul_services_max_fails(ngx_conf_t *cf, ngx_command_t *cmd,
+    void *conf)
+{
+    ngx_str_t                         *value;
+    ngx_http_upsync_srv_conf_t        *upscf;
+
+    upscf = ngx_http_conf_get_module_srv_conf(cf, ngx_http_upsync_module);
+    value = cf->args->elts;
+
+    upscf->consul_services_max_fails = ngx_atoi(value[1].data, value[1].len);
+
+    if (upscf->consul_services_max_fails == (ngx_uint_t) NGX_ERROR) {
+        return NGX_CONF_ERROR; 
+    }
+
+    return NGX_CONF_OK;
+}
+
+
+static char *
+ngx_http_upsync_consul_services_fail_timeout(ngx_conf_t *cf, ngx_command_t *cmd,
+    void *conf)
+{
+    ngx_str_t                         *value;
+    ngx_http_upsync_srv_conf_t        *upscf;
+
+    upscf = ngx_http_conf_get_module_srv_conf(cf, ngx_http_upsync_module);
+    value = cf->args->elts;
+
+    upscf->consul_services_fail_timeout = ngx_parse_time(&value[1], 1);
+
+    if (upscf->consul_services_fail_timeout == (time_t) NGX_ERROR) {
+        return NGX_CONF_ERROR; 
+    }
+
+    return NGX_CONF_OK;
+}
+
+
+static char *
+ngx_http_upsync_fallback_peer(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+{
+    ngx_str_t                         *value;
+    ngx_http_upsync_srv_conf_t        *upscf;
+    ngx_addr_t                        *addrs;
+
+    upscf = ngx_http_conf_get_module_srv_conf(cf, ngx_http_upsync_module);
+    value = cf->args->elts;
+
+    addrs = ngx_http_upsync_addrs(cf->pool, value[1].data);
+    if (addrs == NULL) {
+        return NGX_CONF_ERROR;
+    }
+
+    upscf->fallback_peer.sockaddr = addrs->sockaddr;
+    upscf->fallback_peer.socklen = addrs->socklen;
+    upscf->fallback_peer.name = value[1];
+    upscf->fallback_peer.server = value[1];
+    upscf->fallback_peer.current_weight = 1;
+    upscf->fallback_peer.effective_weight = 1;
+    upscf->fallback_peer.weight = 1;
+    upscf->fallback_peer.max_fails = 2;
+    upscf->fallback_peer.fail_timeout = 10;
+    upscf->fallback_peer.down = 1;
+
+    return NGX_CONF_OK;
+}
+
+
 static void
 ngx_http_upsync_process(ngx_http_upsync_server_t *upsync_server)
 {
@@ -963,6 +1065,8 @@ ngx_http_upsync_del_peers(ngx_cycle_t *cycle,
     ngx_http_upstream_rr_peer_t   *del_peer = NULL, *tmp_del_peer = NULL;
     ngx_http_upstream_rr_peers_t  *peers = NULL;
     ngx_http_upstream_srv_conf_t  *uscf;
+    ngx_http_upsync_srv_conf_t    *upscf;
+    ngx_http_upstream_rr_peer_t   *fallback_peer = NULL;
 
     len = sizeof(struct sockaddr);
     uscf = upsync_server->uscf;
@@ -981,14 +1085,40 @@ ngx_http_upsync_del_peers(ngx_cycle_t *cycle,
     if (uscf->peer.data == NULL) {
         return NGX_ERROR;
     }
-    
+
+    upscf = upsync_server->upscf;
     peers = (ngx_http_upstream_rr_peers_t *)uscf->peer.data;
 
     if (peers->number <= servers->nelts) {
-        ngx_log_error(NGX_LOG_ERR, cycle->log, 0,
-                      "upsync_del_peer: upstream \"%V\" cannot delete all peers", 
-                      &uscf->host);
-        return NGX_ERROR;
+        if (upscf->fallback_peer.down) {
+            fallback_peer = ngx_calloc(sizeof(ngx_http_upstream_rr_peer_t),
+                                       cycle->log);
+            if (fallback_peer == NULL) {
+                return NGX_ERROR;
+            }
+            *fallback_peer = upscf->fallback_peer;
+            fallback_peer->sockaddr = ngx_calloc(sizeof(struct sockaddr_in),
+                                                 cycle->log);
+            if (fallback_peer->sockaddr == NULL) {
+                ngx_free(fallback_peer);
+                return NGX_ERROR;
+            }
+            *fallback_peer->sockaddr = *upscf->fallback_peer.sockaddr;
+            fallback_peer->name.data = ngx_alloc(fallback_peer->name.len,
+                                                 cycle->log);
+            if (fallback_peer->sockaddr == NULL) {
+                ngx_free(fallback_peer);
+                ngx_free(fallback_peer->sockaddr);
+                return NGX_ERROR;
+            }
+            ngx_memcpy(fallback_peer->name.data, upscf->fallback_peer.name.data,
+                       fallback_peer->name.len);
+        } else {
+            ngx_log_error(NGX_LOG_ERR, cycle->log, 0,
+                          "upsync_del_peer: upstream \"%V\" cannot delete all peers",
+                          &uscf->host);
+            return NGX_ERROR;
+        }
     }
 
     n = peers->number - servers->nelts;
@@ -1042,6 +1172,14 @@ ngx_http_upsync_del_peers(ngx_cycle_t *cycle,
     peers->weighted = (w != n);
     peers->total_weight = w;
 
+    if (fallback_peer != NULL) {
+        peers->peer = fallback_peer;
+        peers->single = 1;
+        peers->number = 1;
+        peers->weighted = 0;
+        peers->total_weight = 1;
+    }
+
     if (upsync_server->upscf->upsync_lb == NGX_HTTP_LB_HASH_KETAMA) {
         ngx_http_upsync_del_chash_peer(uscf);
     }
@@ -1441,8 +1579,8 @@ ngx_http_upsync_consul_services_parse_json(void *data)
 
         /* default value, server attribute */
         upstream_conf->weight = 1;
-        upstream_conf->max_fails = 2;
-        upstream_conf->fail_timeout = 10;
+        upstream_conf->max_fails = upsync_server->upscf->consul_services_max_fails;
+        upstream_conf->fail_timeout = upsync_server->upscf->consul_services_fail_timeout;
 
         upstream_conf->down = 0;
         upstream_conf->backup = 0;
@@ -1986,6 +2124,11 @@ ngx_http_upsync_create_srv_conf(ngx_conf_t *cf)
 
     upscf->strong_dependency = NGX_CONF_UNSET_UINT;
 
+    upscf->consul_services_max_fails = NGX_CONF_UNSET_UINT;
+    upscf->consul_services_fail_timeout = NGX_CONF_UNSET;
+
+    ngx_memzero(&upscf->fallback_peer, sizeof(ngx_http_upstream_rr_peer_t));
+
     upscf->conf_file = NGX_CONF_UNSET_PTR;
 
     upscf->upsync_type_conf = NGX_CONF_UNSET_PTR;
@@ -2037,6 +2180,14 @@ ngx_http_upsync_init_srv_conf(ngx_conf_t *cf, void *conf, ngx_uint_t num)
         upscf->strong_dependency = 0;
     }
 
+    if (upscf->consul_services_max_fails == NGX_CONF_UNSET_UINT) {
+        upscf->consul_services_max_fails = 2;
+    }
+
+    if (upscf->consul_services_fail_timeout == NGX_CONF_UNSET) {
+        upscf->consul_services_fail_timeout = 10;
+    }
+
     if (upscf->upsync_dump_path.len == NGX_CONF_UNSET_SIZE) {
         buf = ngx_pcalloc(cf->pool, 
                           ngx_strlen("/tmp/servers_.conf") + uscf->host.len + 1);
```

gfrankliu commented:
How does upsync_fallback_peer work? Is it only for "consul_service" type, or will it also work for old "consul" type?
When a backend service fails a consul health check, it will become invisible to nginx. When all backend services are gone, I assume the upsync_ballback_peer will start to take traffic? When the backend service is back, will you remove the upsync_ballback_peer ?

lyokha commented:
It should work for any type of service (including even etcd, but we tested only for *consul_services*). No, it doesn't accept traffic when all peers fail, actually it's getting statically marked as *down* at the early configuration read phase. This fallback peer is only a workaround for upsync inability of removing all peers (it prints ``cannot delete all peers`` in this case and won't delete anything). In our cases we have situations when we have to delete all servers in a service (when it just do not have servers in its list) - in this case instead of just letting nginx use the old peers that must have been deleted, we really remove them from peers and substitute there our *fallback* peer marked as *down*. Now nginx returns 502 that shows well that the service is down, and as soon as consul reports that new servers are on, the fallback peer gets removed (using normal upsync mechanism for deletion peers). It may repeat many times: the algorithm is crucial for us because we must ensure that when a service has gone, its peers must not be present in the nginx upstream list.

xiaokai-wang commented:
@lyokha I make some comment in your patch. Can you have a look? Thanks.

lyokha commented:
@xiaokai-wang Where do I look at them?

xiaokai-wang commented:
@lyokha  https://github.com/weibocom/nginx-upsync-module/pull/148/files

lyokha commented:
@xiaokai-wang Hmm, by some reason I see no comments there. Probably I have to enable them by clicking some button on the page?

lyokha commented on src/ngx_http_upsync_module.c:
```diff
@@ -1353,6 +1366,148 @@ ngx_http_upsync_consul_parse_json(void *data)
+
+    if (ngx_array_init(&ctx->upstream_conf, ctx->pool, 16,
+                       sizeof(*upstream_conf)) != NGX_OK)
+    {
```
In function *ngx_http_upsync_consul_parse_json()* it's put this way: actually I copied this code from this function. So do you really want that I change that?

lyokha commented on src/ngx_http_upsync_module.c:
```diff
@@ -1353,6 +1366,148 @@ ngx_http_upsync_consul_parse_json(void *data)
+        }
+
+        upstream_conf = ngx_array_push(&ctx->upstream_conf);
+        if (upstream_conf == NULL) {
```
my bad, I will fix it

lyokha commented on src/ngx_http_upsync_module.c:
```diff
@@ -1353,6 +1366,148 @@ ngx_http_upsync_consul_parse_json(void *data)
+
+                if (attr_value == NGX_ERROR || attr_value <= 0) {
+                    ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, 0,
+                          "upsync_parse_json: \"weight\" value is invalid"
```
Ok

lyokha commented on src/ngx_http_upsync_module.c:
```diff
@@ -1353,6 +1366,148 @@ ngx_http_upsync_consul_parse_json(void *data)
+                } else {
+                    upstream_conf->weight = attr_value;
+                }
+            }
```
Actually it gets set on every use in every iteration, so it should be fine.

lyokha commented on README.md:
```diff
@@ -260,7 +260,19 @@ show all upstreams.
+
+You can add or delete backend server through consul_ui or http_interface. Below are examples for key/value store.
 
 http_interface example:
```
Here are examples about addition and deletion KV. If I understand this well, catalogue of services is mostly about static configuration in consul configs. Additionally, I do not know consul interface very well, so at this moment I do not know what examples should be added here.

lyokha commented:
@xiaokai-wang See my comments, especially on `{` indentation and `attr_value`: each use of `arttr_value` is always accompanied by setting it first, so I'm not sure if it makes sense to reset it (if so then where?)

xiaokai-wang commented:
Thanks @lyokha .

Dont see a choice for squeezing the commits to one. It's a little strange. Let to be.

Thanks again.

xiaokai-wang commented:
@lyokha It's great. Please make a new pr. I hope it's optional, have no effect on others.
Sorry to reply late.

gfrankliu commented:
upsync_fallback_peer is a good idea but shouldn't "fake server" act like fallback when there is no servers from consul?

Based on https://github.com/weibocom/nginx-upsync-module/blob/master/README.md
```
        # fake server otherwise ngx_http_upstream will report error when startup
        server 127.0.0.1:11111;
```
During my tests, during nginx starts if upsync couldn't get the peers from consul (e.g.: consul is down), the "fake server" is used for the upstream. Can't we do the same thing if you remove all peers from consul? upsync can use the "fake server" instead of reporting `cannot delete all peers`

lyokha commented:
@gfrankliu Yes, it looks like the best choice, but in this case we should grant this fake server semantics, that it does not have at the moment: the user must know how this *technically inevitable* (which is its only semantics right now) server will now be used. This does not look like a big problem at the first glance: only README update. But what to do when user provides multiple servers, valid and responsive servers etc.? I would just put into peers all statically defined servers with all their statically defined parameters when there is no server from consul and remove them when consul reports new servers. What you think about this?

gfrankliu commented:
Sounds good to me. I thought that is how it is done today already. During my test when consul was failing or returning no peers during nginx startup, I only saw the fake server in the peer list. As you suggested, we just need to do the same when we remove all peers from consul AFTER nginx already running.

gfrankliu commented:
People can put "down" parameter if those fake servers are not meant to be used, or people can put real fallback server here when consul returns nothing.
